In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv('./OnlineRetail.csv', encoding= 'ISO-8859-1')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [10]:
#Question1: The company sells products from how many countries?
country = df.Country.unique()
len(country)

38

The company sells products of 38 countries in the world


In [13]:
#Question2:How many orders have been sold and what is the total revenue?
def cal_revenue(df):
    price, quantinty = df['UnitPrice'],df['Quantity']
    return np.sum(price*quantinty)

In [14]:
total_revenue = cal_revenue(df)
total_revenue

9747747.933999998

In [15]:
orders = df.InvoiceNo.unique()
len(orders)

25900

- The number of orders are 25900
- The total of revenue is 9747747.934$



In [19]:
#Question3:Top 10 best seller products
top_10_product = df.groupby(by =['StockCode','Description']).agg({'Quantity':sum}).reset_index()
top_10_product = top_10_product.sort_values('Quantity',ascending = False)
top_10_product.head(10)

StockCode                         Description  Quantity
3617     84077   WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
4197    85099B             JUMBO BAG RED RETROSPOT     47363
3956     84879       ASSORTED COLOUR BIRD ORNAMENT     36381
1333     22197                      POPCORN HOLDER     36334
531      21212     PACK OF 72 RETROSPOT CAKE CASES     36039
4223    85123A  WHITE HANGING HEART T-LIGHT HOLDER     35025
2344     23084                  RABBIT NIGHT LIGHT     30680
1630     22492             MINI PAINT SET VINTAGE      26437
1775     22616          PACK OF 12 LONDON TISSUES      26315
1122     21977  PACK OF 60 PINK PAISLEY CAKE CASES     24753

In [20]:
#Question4:Top 10 products have the highest revenue?
products = df.groupby(by= ['StockCode','Description'])

In [21]:
products_revenue = products.apply(cal_revenue).reset_index(name = 'Total_revenue')
products_revenue

StockCode                          Description  Total_revenue
0            10002          INFLATABLE POLITICAL GLOBE          759.89
1            10080             GROOVY CACTUS INFLATABLE         119.09
2            10080                                check           0.00
3            10120                         DOGGY RUBBER          40.53
4           10123C                HEARTS WRAPPING TAPE            3.25
...            ...                                  ...            ...
4787  gift_0001_20  to push order througha s stock was            0.00
4788  gift_0001_30   Dotcomgiftshop Gift Voucher £30.00         175.53
4789  gift_0001_40   Dotcomgiftshop Gift Voucher £40.00         100.70
4790  gift_0001_50   Dotcomgiftshop Gift Voucher £50.00         167.56
4791             m                               Manual           2.55

[4792 rows x 3 columns]

In [22]:
products_revenue = products_revenue.sort_values('Total_revenue', ascending= False)
products_revenue.head(10)

StockCode                         Description  Total_revenue
4780       DOT                      DOTCOM POSTAGE      206245.48
1554     22423            REGENCY CAKESTAND 3 TIER      164762.19
3295     47566                       PARTY BUNTING       98302.98
4223    85123A  WHITE HANGING HEART T-LIGHT HOLDER       97715.99
4197    85099B             JUMBO BAG RED RETROSPOT       92356.03
2344     23084                  RABBIT NIGHT LIGHT       66756.59
4783      POST                             POSTAGE       66230.64
1207     22086     PAPER CHAIN KIT 50'S CHRISTMAS        63791.94
3956     84879       ASSORTED COLOUR BIRD ORNAMENT       58959.73
3525     79321                       CHILLI LIGHTS       53768.06